### PR TITLE
Three in one challenge Scrive

### PR DESCRIPTION
**Pull request by Dusste**
Hi,
first run: `npm i` then `npm start`. 
In second tab in terminal please run  server that simulates REST api: 
`json-server --watch server/server.json -p 5010 --cors --delay 1500`

Reason why I have committed as different contributor is that I have SSH key of my old Github account on my laptop.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -5,6 +5,9 @@
   <meta name="viewport" content="width=device-width, initial-scale=1.0">
   <link media="all" rel="stylesheet" href="src/styles.css">
   <title>Elm challenge</title>
+  <!-- <link rel="preconnect" href="https://fonts.googleapis.com">
+  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin> -->
+  <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@100;200;300;400;500;600;800&display=swap" rel="stylesheet">
```
Comment (kraklin) [inline index.html:10]: I really like that you used the same font that we are officially using in Scrive. Really nice touch ❤️

File: public/logo.png
Comment (kraklin) [inline public/logo.png:1]: Again, really nice touch with trying to go with style of an actual Scrive app, logo included. ❤️

File: src/Util.elm [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            "Request failed with status code: " ++ String.fromInt statusCode
+
+        Http.BadBody message ->
+            message
+
+
+
+{-
+   - Email parser
+   -- valid email should contain valid parts, separated by "@", "."
```
Comment (kraklin) [inline src/Util.elm:29]: NIT: email without domain is also valid :D e.g. `tom@localhost` is a valid email.

File: src/Util.elm [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                        if String.isEmpty afterDot || not (isLowerCase afterDot) then
+                            P.problem "Invalid email"
+
+                        else
+                            P.succeed afterDot
+                    )
+           )
+
+
+emailParser : Parser ConstructEmail
```
Comment (kraklin) [inline src/Util.elm:90]: Q: Why did you went with using parser for email validation? What are the benefits of this option for you?

File: src/Util.elm [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+        Ok _ ->
+            Ok (Email email)
+
+
+
+{-
+   - Phone parser
+   -- "+" sign on beggining
+   -- "12345678910" 11 digits string after
```
Comment (kraklin) [inline src/Util.elm:115]: NIT: sometimes there can be less or more numbers in the phone number. If that would be a requirement, would you still use parser or something else? Why?

File: src/Util.elm [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+
+beforeEtParser : Parser String
+beforeEtParser =
+    P.succeed (\s -> s)
+        |= (P.chompWhile Char.isAlphaNum
+                |> P.getChompedString
+                |> P.andThen
+                    (\beforeEt ->
+                        if String.isEmpty beforeEt || not (isLowerCase beforeEt) then
```
Comment (kraklin) [inline src/Util.elm:49]: Q: why did you add the `isLowerCase` condition? emails can contain upper case as well.

File: src/Util.elm (hunk 2 above)
Comment (kraklin) [reply src/Util.elm:29]: Q: Emails are hard to validate because they can be complex. Would your parser handle e.g. `tom+test@my-domain.brand`?

File: src/Util.elm [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    = Email String
+
+
+type alias ConstructEmail =
+    { beforeEt : String, afterEt : String, afterDot : String }
+
+
+beforeEtParser : Parser String
+beforeEtParser =
+    P.succeed (\s -> s)
```
Comment (kraklin) [inline src/Util.elm:44]: NIT: I would use `identity` here. What are the benefits of using `(\s -> s)` instead?

File: src/styles.css [hunk 7]
```diff
@@ -1,3 +1,43 @@
...
+  30% {transform: translate(-50% , -50%) rotate(0deg)}
+  50% {transform: translate(-50% , -50%) rotate(180deg)}
+  75% , 100% {transform: translate(0, 0) rotate(180deg)}
+}
+@keyframes mvrx {
+  0% , 10%  {transform: translate(0 , 0) rotate(0deg)}
+  30% {transform: translate(50% , 50%) rotate(0deg)}
+  50% {transform: translate(50% , 50%) rotate(180deg)}
+  75% , 100% {transform: translate(0, 0) rotate(180deg)}
+}
```
Comment (kraklin) [inline src/styles.css:43]: Love the animation ❤️

File: src/Main.elm [hunk 8]
```diff
@@ -1,22 +1,140 @@
...
+
+-- import Data
 
 import Browser
-import Data
-import Html exposing (Html)
-import Html.Attributes as Attrs
+import Data exposing (userGroup)
+import Dict exposing (Dict)
+import Html exposing (Html, text)
```
Comment (kraklin) [inline src/Main.elm:8]: Seems like you are using qualified imports for every element but `text`. Why did you not use `Html.text` and you chose to expose only this one?

File: src/Main.elm [hunk 9]
```diff
@@ -1,22 +1,140 @@
...
+import Html.Attributes as Attr
+import Html.Events exposing (onCheck, onClick, onInput, onMouseEnter, onMouseLeave)
```
Comment (kraklin) [inline src/Main.elm:10]: Q: Little bit of inconsistency for me again - why do you chose to alias Attributes but have expose of Events? Why not to use `import Html.Events as Events` instead and then have `Events.onClick`?

File: src/Main.elm [hunk 10]
```diff
@@ -1,22 +1,140 @@
...
 
 
 ---- MODEL ----
 
 
+type Data
+    = Initial
+    | Loading
+    | Error String
+    | Success UserGroup
```
Comment (kraklin) [inline src/Main.elm:34]: This seems to be exactly what [RemoteData](https://package.elm-lang.org/packages/krisajenkins/remotedata/latest/RemoteData#RemoteData) looks like. Why did you use your solution instead of using that package?

File: src/Main.elm [hunk 11]
```diff
@@ -1,22 +1,140 @@
...
+    , showTooltip : Bool
+    , toggleDropdown : Bool
+    , errors : Set String
+    , formData : FormData
+    , isFormSent : Bool
+    }
+
+
+type alias FormData =
+    { inputFields : Dict String String
```
Comment (kraklin) [inline src/Main.elm:55]: I really like the usage of `Dict` for input fields ❤️

File: src/Main.elm [hunk 12]
```diff
@@ -25,40 +143,939 @@ init =
...
+                    settings
+
+                { email, phone, companyName, address, zip, city, country, preferredContactMethod } =
+                    contactDetails
+
+                { formData } =
+                    model
+
+                fillInputFields =
+                    Dict.fromList [ ( "preparation", preparation ), ( "closed", closed ), ( "canceled", canceled ), ( "timedOut", timedOut ), ( "rejected", rejected ), ( "error", error ), ( "email", email ), ( "phone", phone ), ( "companyName", companyName ), ( "address", address ), ( "zip", zip ), ( "city", city ), ( "country", country ) ]
```
Comment (kraklin) [inline src/Main.elm:188]: NIT: I would probably prefer this to be formatted with each option at separate line, just for readability.

Comment (kraklin) [inline src/Main.elm:188]: Q: I am little bit afraid that with this approach of having the key as arbitrary string we would be prone to error with misspelling the key when we would like to call it. What are the upsides of using String as a key here? How would you make this safe?

File: src/Main.elm [hunk 13]
```diff
@@ -25,40 +143,939 @@ init =
...
+                        , tags = tags
+                        , prefMethod = preferredContactMethod
+                    }
+            in
+            ( { model | userGroup = Success userGroup, formData = updateFormData }, Cmd.none )
+
+        ActiveTab tab ->
+            ( { model | tabActive = tab }, Cmd.none )
+
+        ShowTooltip ->
```
Comment (kraklin) [inline src/Main.elm:203]: NIT: The name is misleading - it does not always show the tooltip. I would call it ToggleTooltip or something similar.

File: src/Main.elm [hunk 14]
```diff
@@ -25,40 +143,939 @@ init =
...
+                                let
+                                    errorMsg =
+                                        "Details Tab: Invalid email"
+                                in
+                                case Util.parseEmail (inputFields |> Dict.get "email" |> Maybe.withDefault "") of
+                                    Ok _ ->
+                                        Set.remove errorMsg Set.empty
+
+                                    Err _ ->
+                                        Set.insert errorMsg Set.empty
```
Comment (kraklin) [inline src/Main.elm:398]: NIT: This seems like a really deeply nested let..in blocks, which is hard to keep track of what context are you in. I would probably prefer the to be chunked and refactored away

File: src/Main.elm [hunk 15]
```diff
@@ -25,40 +143,939 @@ init =
...
+                                    Phone ->
+                                        let
+                                            errorMsg =
+                                                "Details Tab: Phone is mandatory field"
+                                        in
+                                        if String.isEmpty phoneValue then
+                                            Set.insert errorMsg Set.empty
+
+                                        else
+                                            Set.remove errorMsg Set.empty
```
Comment (kraklin) [inline src/Main.elm:447]: Q: this seems to be repeating quite a lot - what would be a good reason to keep those insert/remove inlined and not refactored/abstracted away

File: src/Main.elm [hunk 16]
```diff
@@ -25,40 +143,939 @@ init =
...
+decodeUserGroup =
+    Decode.succeed UserGroup
+        |> required "settings" decodeSettings
+        |> required "contact_details" decodeContactDetails
+        |> required "tags" decodeTags
+
+
+decodeSettings : Decoder Settings
+decodeSettings =
+    Decode.succeed Settings
```
Comment (kraklin) [inline src/Main.elm:495]: Q: This approach suffers from being error prone to parse wrong value, if you e.g. swap lines 499 and 500 below. What would you do to minimize the risk of swapping those values?

File: src/Main.elm [hunk 17]
```diff
@@ -25,40 +143,939 @@ init =
...
-        [ header "Let's start your task"
-        , subheader "Here are your data:"
-        , Html.pre [ Attrs.class "my-8 py-4 px-12 text-sm bg-slate-100 font-mono shadow rounded" ] [ Html.text Data.userGroup ]
-        , header "Now turn them into form."
-        , subheader "See README for details of the task. Good luck 🍀 "
+view model =
+    Html.div [ Attr.class "flex flex-col sm:w-full md:w-[626px] items-center mx-auto" ]
+        [ case model.userGroup of
+            Initial ->
+                text ""
```
Comment (kraklin) [inline src/Main.elm:640]: It seems that Initial is actually never used. Is there any reason to have it or would you rather remove it completely?

File: src/Main.elm [hunk 18]
```diff
@@ -25,40 +143,939 @@ init =
...
+                        , Html.h1 [ Attr.class "text-3xl  md:text-5xl mb-10" ] [ text "User Group Settings Form" ]
+                        ]
+                    , if model.isFormSent then
+                        Html.div [ Attr.class "flex mb-4 text-green-400 rounded border border-green-400 bg-green-200 px-3 py-2" ]
+                            [ Html.span [ Attr.class "mr-2 flex w-[20px]" ] [ Icons.checkIcon ]
+                            , Html.p [] [ text "Success ! Form has been sent !" ]
+                            ]
+
+                      else
+                        text ""
```
Comment (kraklin) [inline src/Main.elm:668]: NIT: Could be better to use [Html.Extra.viewIf](https://package.elm-lang.org/packages/elm-community/html-extra/latest/Html-Extra#viewIf)

File: src/Main.elm [hunk 19]
```diff
@@ -25,40 +143,939 @@ init =
...
+                    , Html.div [ Attr.class "" ]
```
Comment (kraklin) [inline src/Main.elm:669]: Q: what is the meaning of having `Attr.class ""`?

File: src/Main.elm [hunk 20]
```diff
@@ -25,40 +143,939 @@ init =
...
+                                        viewSettings model.formData userGroup.settings.isInherited model.showTooltip
+
+                                    DetailsTab ->
+                                        viewContactDetails model.formData { isInherited = userGroup.contactDetails.isInherited, prefMethod = model.formData.prefMethod } model.toggleDropdown
+
+                                    TagsTab ->
+                                        viewTags model.formData.tags model.formData.inputFields
+                                ]
+                            ]
+                        , Html.button [ onClick <| SubmitForm { isSettingsInherited = userGroup.settings.isInherited, isDetailsInherited = userGroup.contactDetails.isInherited }, Attr.class "flex font-semi-bold px-5 py-2 rounded-full bg-blue-400 hover:bg-blue-300 active:bg-blue-500 easy-in-out transition-all text-white" ] [ Html.span [ Attr.class "mr-2 flex w-[20px]" ] [ Icons.checkIcon ], text "Submit" ]
```
Comment (kraklin) [inline src/Main.elm:687]: NIT: This is too long for my taste it makes things hard to read for me.

File: src/Main.elm [hunk 21]
```diff
@@ -25,40 +143,939 @@ init =
...
+                , requiredField = prefMethod == Email
+                , textValue = "Email"
+                , inputValue = inputFields |> Dict.get "email" |> Maybe.withDefault ""
+                , maxWidthFieldClass = "w-[132px]"
+                , overrideClass =
+                    if prefMethod == Email then
+                        Just "rounded-t-none ring-blue-400 ring-1"
+
+                    else
+                        Nothing
```
Comment (kraklin) [inline src/Main.elm:937]: Q: what is the reason not to refactor this away? Seems like it is duplicated aver and over

**Review by kraklin — COMMENTED** [review]
Hey Dusan,

thank you for your task. I left comments and questions as a comments in the code. NIT are nitpicks - more or less my observations. Q are questions about the choices and decisions - can you please answer them?

Really nice touch with using the font and logo that we are actually using. Also mimicking the button style of the app... really nice detail. ❤️ Also the loader is really awesome.

I just wanted to ask what would be your next steps with the code if we would like to have it accessible and responsive. (you don't need to code it, just describe the process please :D)

Overall I like this solution 👍 Well done.

File: src/Util.elm (hunk 2 above)
Comment (Dusste) [reply src/Util.elm:29]: It would actually flag an error due to special characters ("-", "+") which should be allowed by usual web applications standards. No excuse for me, but tried to write couple of tests - couldn't find good way to handle opaque type as a return type

Comment (Dusste) [reply src/Util.elm:29]: > NIT: email without domain is also valid :D e.g. `tom@localhost` is a valid email.

True, I was focused more on a conventional company-standard way with something@something.com, but yeah, good point definitely 👍

File: src/Util.elm (hunk 3 above)
Comment (Dusste) [reply src/Util.elm:90]: No obvious benefits. I am slightly more familiar with writing parsers then writing regular expressions :) Between copy/paste regex from the internet or going towards bullet-proof package (which would be my go-to option in real-world) I took decision to go with a custom solution, which even with flaws makes sense when writing a tech task IMO

File: src/Util.elm (hunk 4 above)
Comment (Dusste) [reply src/Util.elm:115]: Yeah, true, requirements wasn't too specific, so i saw what could come from Scrive API documentation. Agree, in real world phone numbers comes in many forms and sizes, I guess than it would be up to us to set some rules and boundaries when we write our application. That would be my answer: If we would have hard requirements, let's say: plus sign then country code then phone blah blah -> Would go with parser, if its very rudimental: "+" and any amount of numbers, think regex can work easily.

File: src/Main.elm (hunk 17 above)
Comment (Dusste) [reply src/Main.elm:640]: Given that "get" request is triggering on init, it makes a lot of sense to remove it completely 👍

File: src/Main.elm (hunk 18 above)
Comment (Dusste) [reply src/Main.elm:668]: Oh that is very useful package ! Thanks for bringing that one, never used it or heard about it 🤓 Completely agree

File: src/Main.elm (hunk 19 above)
Comment (Dusste) [reply src/Main.elm:669]: It was leftover from  previous attempt to style that element, then eventually styled parent or child, I think there are at least 2 more occasions where i forgot to delete unused Attr.class ""

File: src/Main.elm (hunk 20 above)
Comment (Dusste) [reply src/Main.elm:687]: Would it makes sense to "break" it in next line and let elm format do the job ? Or is it more in terms of a size of a record + gigantic chained tailwind class ?

File: src/Main.elm (hunk 21 above)
Comment (Dusste) [reply src/Main.elm:937]: I don't think 3 conditionals are duplicate, I need those overrides only in specific cases, and because **viewFieldset** is handling all inputs, needed the way to distinguish them.

File: src/Util.elm (hunk 5 above)
Comment (Dusste) [reply src/Util.elm:49]: Can't come up to a solid answer, for some reason I added that, I think I am not a big fan of adding uppercase letters in a email :) but never revisited implementation afterwords. True, emails can contain uppercase letters.

File: src/Util.elm (hunk 6 above)
Comment (Dusste) [reply src/Util.elm:44]: In my case - it just worked, didn't bother to find better solution, so I moved on. I guess identity would be cleaner and more idiomatic way definitely 👍

File: src/styles.css (hunk 7 above)
Comment (Dusste) [reply src/styles.css:43]: It was copy/pasted from somewhere, can't take credits for this css witchcrafts :) Was a bit inspired with your Rubiks cube tbh

File: src/Main.elm (hunk 8 above)
Comment (Dusste) [reply src/Main.elm:8]: haha that is one of my "rules" how I write things in Elm. For example, for Html, I would exposed "text" and "Html" type (I hate to see Html.Html), for Parser it would be those crazy characters (|. , |=) I tend to use qualified imports with certain amount of exposed stuff, that can't be in conflict with any other package/module. So definitely I would exposed at least types, for packages.

File: src/Main.elm (hunk 9 above)
Comment (Dusste) [reply src/Main.elm:10]: It was just force of habit: exposing methods from Events but using alias for Attributes. Can't explain it, I think its just side effect of writing Elm applications as a one-man-band. In case of writing production code I would definitely be more focused on consistency.

File: src/Main.elm (hunk 10 above)
Comment (Dusste) [reply src/Main.elm:34]: I was definitely inspired by the RemoteData package. Haven't seen benefit in including whole package if I needed only one custom type. It's like that jungle/banana dilemma. I wanted only 🍌

File: src/Main.elm (hunk 11 above)
Comment (Dusste) [reply src/Main.elm:55]: Thanks, I was a bit concerned about performance, tried to dig a but of informations about the topic, eventually did it with Dict given that number of inputs was like 13 or something. Not sure if I would gone same route if number was 20k or something but that is close to impossible at least for input fields.

File: src/Main.elm (hunk 12 above)
Comment (Dusste) [reply src/Main.elm:188]: Yeah, It was stupid decision from my side (here and in some other places) in order to reduce number of lines :( Pretty noob concerns from my side

Comment (Dusste) [reply src/Main.elm:188]: Agree, error prone. Given that Dict key can be only comparable, so can't go away from string BUT I have `fromIdToString` function https://github.com/Dusste/elm-challenge/blob/three-in-one-challenge-scrive-dusste/src/Main.elm#L582 already there, so  Dict.fromList [ ( ((fromIdToString PrepId), preparation ) ...
i already used it to convert Id to string in some places, might forgot about this case or it felt cumbersome.

File: src/Main.elm (hunk 13 above)
Comment (Dusste) [reply src/Main.elm:203]: `Toggle` is my usual go to ! I already have `ToggleDropdown`

File: src/Main.elm (hunk 14 above)
Comment (Dusste) [reply src/Main.elm:398]: Yeah, it was a very annoying part to validate all the fields in one place. Took that decision so I don't do validation onChange per a field, since in that case (I think) I would need to write all OnChange updates for each fields. So took "less evil" path.

File: src/Main.elm (hunk 15 above)
Comment (Dusste) [reply src/Main.elm:447]: Can't say, didn't know better, maybe just inexperienced way of writing

File: src/Main.elm (hunk 16 above)
Comment (Dusste) [reply src/Main.elm:495]: Yeah, its error prone since its based on a order of fields.  I think `oneOf` would do a trick ?

**Comment by Dusste** [comment]
> Hey Dusan,
> 
> thank you for your task. I left comments and questions as a comments in the code. NIT are nitpicks - more or less my observations. Q are questions about the choices and decisions - can you please answer them?
> 
> Really nice touch with using the font and logo that we are actually using. Also mimicking the button style of the app... really nice detail. ❤️ Also the loader is really awesome.
> 
> I just wanted to ask what would be your next steps with the code if we would like to have it accessible and responsive. (you don't need to code it, just describe the process please :D)
> 
> Overall I like this solution 👍 Well done.
> 

Thank you for your comments and observations, they are very much appreciated and very useful ! 
Regarding question for responsiveness: Strange, I made mobile styles, and tested it on a mobile's browsers simulator, I think it is compatible with mobile and it was looking good. Maybe button for submit should have taken maximum width on mobile width ? 
Regarding accessibility:  
I think form is following that approach (form -> fieldset -> label with "for" -> input with "id") 
Section tag was used, didn't have chance to use "header" or "article" or alt on img tag. TBH don't have that much experience dealing with accessibility throughout my career, I know some best practises and thats it. Can't give any good answer on your question how can I improved at this point, maybe after FrontedMasters course on a topic :P

Thanks for a chance, it was fun :)
